src/solvers/lowering/byte_operators.cpp

```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+        src.id(),
+        index_exprt(src.op(), where),
+        from_integer(0, src.offset().type()),
+        byte_extract_exprt(
+          extract_opcode,
+          value_as_byte_array,
+          from_integer(offset, src.offset().type()),
+          array_typet(unsignedbv_typet(8), subtype_size))),
+      ns);
+
+    result.add_to_operands(std::move(where), std::move(element));
+  }
 
-  if(src.op().type().id()==ID_array)
+  // do the tail
+  if(offset < value_as_byte_array.operands().size())
   {
-    const array_typet &array_type=to_array_type(src.op().type());
-    const typet &subtype=array_type.subtype();
+    const std::size_t tail_size =
+      value_as_byte_array.operands().size() - offset;
+
+    exprt where = simplify_expr(
+      plus_exprt(first_index, from_integer(i, first_index.type())), ns);
+
+    exprt element = lower_byte_operators(
+      byte_update_exprt(
+        src.id(),
+        index_exprt(src.op(), where),
+        from_integer(0, src.offset().type()),
+        byte_extract_exprt(
+          extract_opcode,
+          value_as_byte_array,
+          from_integer(offset, src.offset().type()),
+          array_typet(
+            unsignedbv_typet(8), from_integer(tail_size, size_type())))),
+      ns);
+
+    result.add_to_operands(std::move(where), std::move(element));
+  }
+
+  return simplify_expr(result, ns);
+}
+
+static exprt lower_byte_update_array_vector(
+  const byte_update_exprt &src,
+  const typet &subtype,
+  const array_exprt &value_as_byte_array,
+  const optionalt<exprt> &non_const_update_bound,
+  const namespacet &ns)
+{
+  const bool is_array = src.type().id() == ID_array;
+  exprt size;
+  if(is_array)
+    size = to_array_type(src.type()).size();
+  else
+    size = to_vector_type(src.type()).size();
 
-    // array of bitvectors?
-    if(subtype.id()==ID_unsignedbv ||
-       subtype.id()==ID_signedbv ||
-       subtype.id()==ID_floatbv ||
-       subtype.id()==ID_c_bool ||
-       subtype.id()==ID_pointer)
+  auto subtype_bits = pointer_offset_bits(subtype, ns);
+
+  // fall back to bytewise updates in all non-constant or dubious cases
+  if(
+    !size.is_constant() || !src.offset().is_constant() ||
+    !subtype_bits.has_value() || *subtype_bits == 0 || *subtype_bits % 8 != 0 ||
+    non_const_update_bound.has_value())
+  {
+    return lower_byte_update_array_vector_non_const(
+      src, subtype, value_as_byte_array, non_const_update_bound, ns);
+  }
+
+  std::size_t num_elements =
+    numeric_cast_v<std::size_t>(to_constant_expr(size));
+  mp_integer offset_bytes =
+    numeric_cast_v<mp_integer>(to_constant_expr(src.offset()));
+
+  exprt::operandst elements;
+  elements.reserve(num_elements);
+
+  for(std::size_t i = 0; i < num_elements; ++i)
+  {
+    index_exprt index(src.op(), from_integer(i, index_type()));
+
+    if(
+      (offset_bytes + value_as_byte_array.operands().size()) * 8 <=
+        i * *subtype_bits ||
+      offset_bytes * 8 >= (i + 1) * *subtype_bits)
```

Comment:
Part of the computation could be taken out of the loop and it looks like the loop should be split to avoid unnecessary checks: once the first condition is true then it will always be true?

Reply:
Indeed, I have split the loop into three chunks.